addons/JobChange/jobchange.lua

```diff
@@ -0,0 +1,145 @@
+    local args = L{...}
+    local job = ''
+    if args[1] then 
+        job = args[1]:lower()
+    end
+    local main_sub = ''
+    if command:lower() == 'main' then
+        main_sub = 'main'
+    elseif command:lower() == 'sub' then
+        main_sub = 'sub'
+    elseif command:lower() == 'reset' then
+        windower.add_to_chat(4,"JobChange: Resetting Job")
+        main_sub = 'sub'
+        job = windower.ffxi.get_player().sub_job:lower()
+    else
+        windower.add_to_chat(4,"JobChange Syntax: //jc main|sub JOB  -- Chnages main or sub to target JOB")
+        windower.add_to_chat(4,"JobChange Syntax: //jc reset -- Resets Current Job")
+        return
+    end
+    local conflict = find_conflict(job,self)
+    local jobid = find_job(job,self)
+    if jobid then 
+        local npc = find_job_change_npc()
+        if npc then
+            if not conflict then 
+                jobchange(jobid,main_sub)
+            else
+                local temp_job = find_temp_job(self)            
+                windower.add_to_chat(4,"JobChange: Conflict with "..conflict)
+                if main_sub == conflict then 
+                    jobchange(temp_job,main_sub)
+                    jobchange(jobid,main_sub)
+                else
+                    jobchange(temp_job,conflict)
+                    jobchange(jobid,main_sub)
+                end
+            end
+        else
+            windower.add_to_chat(4,"JobChange: Not close enough to a Moogle!")
+        end        
+    else
+        windower.add_to_chat(4,"JobChange: Could not change "..command.." to "..job:upper().." ---Mistype|NotUnlocked")
+    end
+    
+end)
+
+function find_conflict(job,self)
+    if self.main_job == job:upper() then
+        return "main"
+    end
+    if self.sub_job == job:upper() then
+        return "sub"
+    end
+end
+
+function find_temp_job(self)
+    local starting_jobs = S { 'WAR', 'MNK', 'WHM', 'BLM', 'RDM', 'THF' } 
+    for i in pairs(starting_jobs) do
+        if not find_conflict(i,self) then 
+            for index,value in pairs(res.jobs) do
+                if value.ens == i then
+                    return index
+                end
+            end
+        end
+    end
+end
+
+function find_job(job,self)
+    local jobLevel = self.jobs[job:upper()]
+    for index,value in pairs(res.jobs) do
+        if value.ens:lower() == job and jobLevel > 0 then 
+            return index
+        end
+    end
+end
+
+
+function find_job_change_npc()
+    local info = windower.ffxi.get_info()
+    if not (info.mog_house or S{'Selbina', 'Mhaura', 'Tavnazian Safehold', 'Nashmau', 'Rabao', 'Kazham', 'Norg'}:contains(res.zones[info.zone].english)) then
+        windower.add_to_chat(4,'JobChange: Not in a zone with a Change NPC')
+        return
+    end
+
+    for i, v in pairs(windower.ffxi.get_mob_array()) do
+        if v.distance < 36 and S{'Moogle', 'Nomad Moogle', 'Green Thumb Moogle'}:contains(v.name) then
```

Comment:
A minor note on using `S{}` here. I do not know how much Lua tries to optimize code, I know it does not do a lot. So it wouldn't surprise me, if it couldn't optimize the set creation out of this. Here is what your code *says* to do:

1. Iterate over every mob in the zone
2. Compare the distance
3. Create a new table, containing three values, `'Moogle'`, `'Nomad Moogle'` and `'Green Thumb Moogle'`
4. Turn that table into a set using `S`
5. Check if the mob's name is in that set

You can see number 3 and 4 are repeated for every mob, which can be up to 2k mobs. But the result of those steps (a set containing the names of moogles) is not reliant on the loop variable, so in a similar move to the `jobLevel` thing I mentioned above, we can move it outside of the loop.

While that will realistically not to impact anyone much, the idea behind this would be to save the work where possible. So you can create the set outside of the loop, then it would only be created once:
```lua
    local moogles = S{'Moogle', 'Nomad Moogle', 'Green Thumb Moogle'}
    for i, v in pairs(windower.ffxi.get_mob_array()) do
        if v.distance < 36 and moogles:contains(v.name) then
            return v
        end
    end
```